narwhal/primary/src/synchronizer.rs

```diff
@@ -1,70 +1,340 @@
+/// - Broadcasting created certificates.
+/// `Synchronizer` contains most of the certificate processing logic in Narwhal.
+pub struct Synchronizer {
+    /// Internal data that are thread safe.
+    inner: Arc<Inner>,
 }
 
 impl Synchronizer {
     pub fn new(
         name: PublicKey,
         committee: SharedCommittee,
         worker_cache: SharedWorkerCache,
+        gc_depth: Round,
         certificate_store: CertificateStore,
         payload_store: Store<(BatchDigest, WorkerId), PayloadToken>,
         tx_certificate_fetcher: Sender<Certificate>,
+        tx_new_certificates: Sender<Certificate>,
+        tx_parents: Sender<(Vec<Certificate>, Round, Epoch)>,
         rx_consensus_round_updates: watch::Receiver<Round>,
+        rx_synchronizer_network: oneshot::Receiver<Network>,
         dag: Option<Arc<Dag>>,
+        metrics: Arc<PrimaryMetrics>,
     ) -> Self {
-        let genesis = Self::make_genesis(&committee.load());
-        Self {
+        let committee: &Committee = &committee.load();
+        let genesis = Self::make_genesis(committee);
+        let highest_processed_round = certificate_store.highest_round_number();
+        let highest_created_certificate = certificate_store.last_round(&name).unwrap();
+        let gc_round = (*rx_consensus_round_updates.borrow()).saturating_sub(gc_depth);
+        let (tx_own_certificate_broadcast, _rx_own_certificate_broadcast) =
+            broadcast::channel(1000);
+        let inner = Arc::new(Inner {
             name,
+            committee: committee.clone(),
             worker_cache,
+            gc_depth,
+            gc_round: AtomicU64::new(gc_round),
+            highest_processed_round: AtomicU64::new(highest_processed_round),
+            highest_received_round: AtomicU64::new(0),
             certificate_store,
             payload_store,
             tx_certificate_fetcher,
-            rx_consensus_round_updates,
+            tx_new_certificates,
+            tx_parents,
+            tx_own_certificate_broadcast: tx_own_certificate_broadcast.clone(),
+            rx_consensus_round_updates: rx_consensus_round_updates.clone(),
             genesis,
             dag,
+            metrics,
+            batch_tasks: Mutex::new(JoinSet::new()),
+            certificate_senders: Mutex::new(JoinSet::new()),
+            certificates_aggregators: Mutex::new(BTreeMap::new()),
+            pending: tokio::sync::Mutex::new(HashMap::new()),
+        });
+
+        // Start a task to recover parent certificates for proposer.
+        let weak_inner = Arc::downgrade(&inner);
+        spawn_monitored_task!(async move {
+            let Some(inner) = weak_inner.upgrade() else {
+                // this happens if Narwhal is shutting down.
+                return;
+            };
+            let last_round_certificates = inner
+                .certificate_store
+                .last_two_rounds_certs()
+                .expect("Failed recovering certificates in primary core");
```

Comment:
Panic is abort in production build: https://github.com/MystenLabs/sui/blob/4baf554f12a3a43368429c2009bd6fd50adb830a/Cargo.toml#L82. Or do you want an explicit abort here? TBH I think with the logic to resend last created headers and certificates on restart, the logic of resending last 2 rounds to proposer seems no longer needed? I can open another PR to discuss this.

Reply:
I believe we still need to keep this logic here as this allow us to propose for the next round - assuming we haven't done already (otherwise the header resend mechanism is the one that supersedes this one).

Reply:
Makes sense!